ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/HD/BackingStore.hs

```diff
@@ -0,0 +1,373 @@
+closeHandlers :: IOLike m => [Handler m ()]
+closeHandlers = [
+    Handler $ \case
+      BS.TVarBackingStoreClosedExn -> pure ()
+      e                            -> throwIO e
+  , Handler $ \case
+      LMDB.DbErrClosed -> pure ()
+      e                -> throwIO e
+  ]
+
+{-------------------------------------------------------------------------------
+  Types under test
+-------------------------------------------------------------------------------}
+
+type T = BackingStoreState K V D
+
+pT :: Proxy T
+pT = Proxy
+
+type K = LedgerTables (SimpleLedgerState (Fixed Word) (Fixed Word)) KeysMK
+type V = LedgerTables (SimpleLedgerState (Fixed Word) (Fixed Word)) ValuesMK
+type D = LedgerTables (SimpleLedgerState (Fixed Word) (Fixed Word)) DiffMK
+
+{-------------------------------------------------------------------------------
+  @'HasOps'@ instances
+-------------------------------------------------------------------------------}
+
+instance Mock.EmptyValues V where
+  emptyValues = polyEmptyLedgerTables
+
+instance Mock.ApplyDiff V D where
+  applyDiff = zipLedgerTables rawApplyDiffs
+
+instance Mock.LookupKeysRange K V where
+  lookupKeysRange = \prev n vs ->
+      case prev of
+        Nothing ->
+          mapLedgerTables (rangeRead n) vs
+        Just ks ->
+          zipLedgerTables (rangeRead' n) ks vs
+    where
+      rangeRead :: Int -> ValuesMK k v -> ValuesMK k v
+      rangeRead n (ApplyValuesMK (DS.Values vs)) =
+        ApplyValuesMK $ DS.Values $ Map.take n vs
+
+      rangeRead' ::
+          Ord k
+        => Int
+        -> KeysMK k v
+        -> ValuesMK k v
+        -> ValuesMK k v
+      rangeRead' n ksmk vsmk =
+          case Set.lookupMax ks of
+            Nothing -> ApplyValuesMK $ DS.Values Map.empty
+            Just  k -> ApplyValuesMK $ DS.Values $
+              Map.take n $ snd $ Map.split k vs
+        where
+          ApplyKeysMK (DS.Keys ks)     = ksmk
+          ApplyValuesMK (DS.Values vs) = vsmk
+
+instance Mock.LookupKeys K V where
+  lookupKeys = zipLedgerTables readKeys
+    where
+      readKeys ::
+          Ord k
+        => KeysMK k v
+        -> ValuesMK k v
+        -> ValuesMK k v
+      readKeys (ApplyKeysMK ks) (ApplyValuesMK vs) =
+        ApplyValuesMK $ DS.restrictValues vs ks
+
+instance Mock.ValuesLength V where
+  valuesLength (SimpleLedgerTables (ApplyValuesMK (DS.Values m))) =
+    Map.size m
+
+instance Mock.MakeDiff V D where
+  diff t1 t2 = zipLedgerTables (rawForgetValues .: rawCalculateDifference) t1 t2
+
+instance Mock.DiffSize D where
+  diffSize (SimpleLedgerTables (ApplyDiffMK (Diff.Diff m))) = Map.size m
+
+instance Mock.KeysSize K where
+  keysSize (SimpleLedgerTables (ApplyKeysMK (Diff.Keys s))) = Set.size s
+
+instance Mock.HasOps K V D
+
+{-------------------------------------------------------------------------------
+  Utilities
+
+  TODO: should become available in a newer version of @quickcheck-dynamic@.
```

Comment:
Maybe it's worth creating a ticket, or submitting a PR to QCD directly.

Reply:
I will create an issue in this repo -- the changes are already present on the tip of `q-d`'s `main` branch, so we should be able to remove these utilities once `q-d-2.1.0` is released. See https://github.com/input-output-hk/ouroboros-network/issues/4228.

Reply:
I've also slightly rephrased the TODO